### PR TITLE
feat(mcp-server): MCP server exposing PilotSwarm sessions, agents, facts, and models

### PR DESCRIPTION
**Pull request by ankitdas-volgapartners**
## Summary

Adds `packages/mcp-server/` — a [Model Context Protocol](https://modelcontextprotocol.io/) server that exposes PilotSwarm sessions, agents, facts, and models to any MCP-compatible client (Claude Desktop, Copilot CLI, Cursor, VS Code, ChatGPT, and others). Both stdio and HTTP transports are supported; HTTP is gated behind a shared-secret header for remote / shared access.

The package is purely additive — nothing in `packages/sdk/`, `packages/portal/`, `packages/cli/`, or `packages/ui-*` is touched. The MCP server creates its own `PilotSwarmClient` internally and connects to the same database a PilotSwarm worker uses.

## Motivation

PilotSwarm already runs sessions and persists state durably; the missing piece is letting external MCP-aware tools drive those sessions, read live facts, manage agents, and inspect model providers without re-implementing the SDK in every client.

This PR brings:

1. **Stdio transport** for local IDE integration (Claude Desktop, Copilot CLI, Cursor, VS Code) — zero-config beyond the standard `.mcp.json`.
2. **HTTP transport** for shared / multi-user access with a `PILOTSWARM_MCP_KEY` shared-secret header so an MCP server can sit on a remote box and serve a team.
3. **First-class MCP surface** for the things PilotSwarm already manages: sessions (list / get / send / messages / stats / archive), agents (list / get / install), facts (read / write / list), models (list providers and models), plus subscriptions for streaming updates.

## What's added

```
packages/mcp-server/
├── bin/                                  CLI entrypoint
├── src/
│   ├── index.ts                          process bootstrap + transport selection
│   ├── server.ts                         MCP server wiring (resources / tools / prompts)
│   ├── resources/
│   │   ├── agents.ts                     list / get installed agents
│   │   ├── facts.ts                      read / list facts
│   │   ├── models.ts                     list model providers + models
│   │   ├── sessions.ts                   list / get sessions + messages
│   │   └── subscriptions.ts              streaming updates surface
│   ├── tools/
│   │   ├── agents.ts                     install / configure agents
│   │   ├── facts.ts                      write / delete facts
│   │   ├── models.ts                     model selection / config
│   │   └── sessions.ts                   send / archive / dehydrate / hydrate
│   └── prompts/skills.ts                 prompt-style skill exposure
├── test-mcp-edge-cases.mjs               edge-case smoke
├── test-mcp-tools.mjs                    tool-surface smoke
├── test-mcp-verify.mjs                   end-to-end verification
├── package.json, tsconfig.json, README.md
```

### Transport modes

| Mode | When to use | Auth |
|---|---|---|
| `stdio` | Local IDE (Claude Desktop, Copilot CLI, Cursor, VS Code) | Process-level |
| `http` | Remote / shared / team | Shared-secret header `PILOTSWARM_MCP_KEY` |

### Surface highlights

- **Resources** are read-only views: sessions, agents, facts, models, subscriptions.
- **Tools** are state-mutating: send a turn into a session, archive a session, write a fact, install an agent, switch model provider.
- **Prompts** expose installed PilotSwarm skills as MCP prompts so a client can route to a named workflow.

## What's *not* changing

- `packages/sdk/`, `packages/portal/`, `packages/cli/`, `packages/ui-core/`, `packages/ui-react/` — untouched.
- No SDK API changes, no schema changes, no CMS schema changes.
- Existing tests under `packages/sdk/test/local/` are unchanged.
- No new top-level scripts; the MCP server is a workspace package with its own `bin/`.
- `pilotswarm-sdk` peer dependency satisfied by current SDK version.

### Stdio (local IDE)

```bash
# .mcp.json (project or ~/.copilot/)
{
  "mcpServers": {
    "pilotswarm": {
      "command": "npx",
      "args": [
        "pilotswarm-mcp",
        "--store", "postgres://...",
        "--model-providers", ".model_providers.json"
      ]
    }
  }
}
```

### HTTP (remote / shared)

```bash
# Server side
PILOTSWARM_MCP_KEY=your-secret-key npx pilotswarm-mcp \
  --transport http --port 3100 \
  --store "$DATABASE_URL" --model-providers .model_providers.json

# Client side — pass the key as a header
curl -H "x-pilotswarm-mcp-key: your-secret-key" http://host:3100/mcp/...
```

`packages/mcp-server/README.md` has Stdio + HTTP `.mcp.json` snippets for Claude Desktop, Copilot CLI, Cursor, VS Code, ChatGPT, and a generic MCP client.

## Testing

- `npm run build` clean (sdk + mcp-server).
- `tsc` clean across the package.
- `test-mcp-tools.mjs` + `test-mcp-edge-cases.mjs` + `test-mcp-verify.mjs` exercise the full tool surface (send, list, archive, fact write/read, agent install) plus edge cases (invalid input shape, missing-session, auth failure on HTTP transport).
- Manually verified with: Claude Desktop (stdio), Copilot CLI (stdio + http), Cursor (stdio).

## Reviewer guidance

Small package, ~1.5k LOC. Suggested review order:

1. `src/server.ts` — MCP server wiring; this is where resources / tools / prompts get registered
2. `src/index.ts` — process bootstrap + transport selection (`--transport stdio|http`)
3. `src/tools/sessions.ts` and `src/tools/facts.ts` — main state-mutating surface; review for input validation + error shape
4. `src/resources/sessions.ts` — read surface most clients exercise first
5. HTTP transport auth path (key header handling) — ensure no bypass
6. `bin/` — CLI argument parsing
7. README.md client snippets — make sure they match what the server actually expects

## Out of scope / follow-ups

- Streaming subscription transport polish (currently single-shot reads)
- Per-client capability negotiation refinement as the MCP spec evolves
- Optional Bearer-token auth for HTTP (currently shared-secret header only)
- Bundling a docker image so non-Node MCP clients have a one-line `docker run`

### CODE REVIEW

File: packages/mcp-server/src/tools/agents.ts [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+            inputSchema: {
+                session_id: z.string().describe("The session to spawn the agent in"),
+                task: z.string().describe("The task description for the new agent"),
+                agent_name: z.string().optional().describe("Optional name for the agent"),
+                model: z.string().optional().describe("Optional model override"),
+            },
+        },
+        async ({ session_id, task, agent_name, model }) => {
+            try {
+                await ctx.mgmt.sendCommand(session_id, {
```
Comment (affandar) [inline packages/mcp-server/src/tools/agents.ts]: Blocking: this enqueues `cmd: "spawn_agent"` through `PilotSwarmManagementClient.sendCommand`, but the durable orchestration command handler does not handle `spawn_agent` management commands. It only handles `set_model`, `list_models`, `get_info`, `done`, `cancel`, and `delete`. The sub-agent `TurnAction` handlers are reached from LLM turn results, not from command events, so this MCP tool returns `{ sent: true }` while the orchestration records `Unknown command: spawn_agent` and no child is spawned.

The same issue affects `message_agent` and `cancel_agent` below. Please either add supported orchestration management commands for these actions and wait/read their command responses, or expose real management APIs for spawn/message/cancel-agent operations and call those here.

File: packages/mcp-server/src/resources/agents.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
+import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
+import type { ServerContext } from "../context.js";
+
+const SYSTEM_AGENTS = ["sweeper", "resourcemgr", "facts-manager"] as const;
```
Comment (affandar) [inline packages/mcp-server/src/resources/agents.ts]: This hardcodes the system agent IDs, so MCP agent resources only exist for `sweeper`, `resourcemgr`, and `facts-manager`. Current PilotSwarm can have other system sessions, and app plugins can add or rename system agents, so this will silently miss valid system agents.

For active system agents, the existing management API already has the data: `ctx.mgmt.listSessions()` returns rows with `isSystem` and `agentId`. Please derive these dynamically and expose `pilotswarm://agents/{agentId}` through a `ResourceTemplate` with a dynamic `list` callback instead of registering a fixed list at startup. If we need configured-but-not-running system agents too, that should be a new management API rather than a hardcoded list here.

File: packages/mcp-server/src/resources/subscriptions.ts [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+                // Session-specific resource
+                const sessionUri = `pilotswarm://sessions/${s.sessionId}`;
+                if (subscriptions.has(sessionUri)) {
+                    try {
+                        await lowLevel.sendResourceUpdated({ uri: sessionUri });
+                    } catch { /* ignore */ }
+                }
+
+                // System agent alias resources
+                if (s.isSystem && s.agentId && SYSTEM_AGENTS.includes(s.agentId)) {
```
Comment (affandar) [inline packages/mcp-server/src/resources/subscriptions.ts]: Same hardcoded-system-agent issue here: resource update notifications are only emitted for the static `SYSTEM_AGENTS` list. If a system agent is renamed or an app-defined system agent is running, subscribers to `pilotswarm://agents/<agentId>` will never get updates.

This can use the session row directly: if `s.isSystem && s.agentId`, emit the `pilotswarm://agents/${s.agentId}` update when subscribed. No static allow-list should be needed for active system sessions.

File: packages/mcp-server/README.md [hunk 4]
```diff
@@ -0,0 +1,386 @@
+# PilotSwarm MCP Server
+
+Exposes PilotSwarm sessions, agents, facts, and models to any MCP-compatible client via the [Model Context Protocol](https://modelcontextprotocol.io/). Connect Claude Desktop, Copilot CLI, Cursor, VS Code, ChatGPT, or any MCP client to a running PilotSwarm instance.
+
+## Quick Start
+
+### Stdio Transport (recommended for local IDEs)
+
+```bash
+npx pilotswarm-mcp --store "$DATABASE_URL" --model-providers .model_providers.json
```
Comment (affandar) [inline packages/mcp-server/README.md]: The quickstart uses `npx pilotswarm-mcp`, but the package name is `pilotswarm-mcp-server` and `pilotswarm-mcp` is only the bin name. A fresh user running this command will have npm try to resolve a package named `pilotswarm-mcp`.

Please change these examples to either `npx pilotswarm-mcp-server ...` if npm resolves the package bin automatically, or the unambiguous form `npx -p pilotswarm-mcp-server pilotswarm-mcp ...`.

File: packages/mcp-server/package.json [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  "bin": {
+    "pilotswarm-mcp": "./dist/bin/pilotswarm-mcp.js"
+  },
+  "exports": {
+    ".": {
+      "import": "./dist/src/index.js",
+      "types": "./dist/src/index.d.ts"
+    }
+  },
+  "scripts": {
```
Comment (affandar) [inline packages/mcp-server/package.json:23]: This PR adds three MCP verification files, but they are not wired into either `./scripts/run-tests.sh` or a package-level test script. `run-tests.sh` only builds/runs Vitest from `packages/sdk/test/local`, and `npm test --workspace=packages/mcp-server` currently fails with `Missing script: "test"`.

Please add an MCP test entry point and wire it into the repo's test path/CI. If these scripts require a live DB/worker, either convert them into deterministic local integration tests with the same setup conventions as `packages/sdk/test/local`, or add explicit `test:mcp:*` scripts and document when CI/release validation runs them.

File: packages/mcp-server/src/tools/models.ts [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+                command: z.string().describe("The command name to send"),
+                args: z
+                    .record(z.string(), z.any())
+                    .optional()
+                    .describe("Optional arguments for the command"),
+            },
+        },
+        async ({ session_id, command, args }) => {
+            try {
+                await ctx.mgmt.sendCommand(session_id, {
```
Comment (affandar) [inline packages/mcp-server/src/tools/models.ts]: This returns `{ sent: true }` for any arbitrary command immediately after enqueueing it, even when the orchestration will reject it as `Unknown command`. That makes MCP clients think commands succeeded when they only reached the queue.

For command-style MCP tools, please wait for/read the command response and return success only when the orchestration reports success; otherwise return an MCP error with the command response error text.

File: packages/mcp-server/README.md [hunk 7]
```diff
@@ -0,0 +1,386 @@
...
+
+### HTTP Transport (recommended for remote/shared access)
+
+```bash
+PILOTSWARM_MCP_KEY=your-secret-key npx pilotswarm-mcp \
+  --transport http --port 3100 \
+  --store "$DATABASE_URL" --model-providers .model_providers.json
+```
+
+> **Prerequisite:** A running PostgreSQL database and a PilotSwarm worker. The MCP server creates a `PilotSwarmClient` internally — it needs the same database the worker connects to.
```
Comment (affandar) [inline packages/mcp-server/README.md:23]: Non-blocking follow-up: the direct `DATABASE_URL` model is acceptable for this initial MCP server, and it matches how the TUI connects today. But please file a tracking issue for exposing the PilotSwarm management/API surface through the portal as HTTP endpoints, then allowing the MCP server to talk to those endpoints instead of requiring direct database access.

Without that, this MCP server will not work in deployments where the only supported interface to the PilotSwarm cluster is the portal. In that topology, users should not need PostgreSQL connectivity or database credentials on the MCP client/server side; they should be able to point the MCP server at the portal/API URL with portal-compatible auth.

File: package-lock.json [hunk 8]
```diff
@@ -3273,6 +3468,12 @@
       "integrity": "sha512-hvpoI6korhJMnej285dSg6nu1+e6uxs7zG3BYAm5byqDsgJNWwxzM6z6iZiAgQR4TJ30JmBTOwqZUw3WlyH3AQ==",
       "license": "MIT"
     },
+    "node_modules/isexe": {
```
Comment (affandar) [inline package-lock.json:3654]: Merge-conflict note: this PR currently conflicts with `main` only in `package-lock.json`. I checked with a temporary worktree merge; the conflict is a generated adjacent-entry collision where this branch adds `node_modules/isexe` and current `main` adds `node_modules/is-wsl` in the same sorted area.

Resolution should be mechanical: merge/rebase onto `main`, regenerate the lockfile from the merged package manifests (`npm install --package-lock-only` or `npm install`), make sure both `isexe` and `is-wsl` entries are retained, then rerun `npm ci` and `npm run build --workspace=packages/mcp-server`.

**Review by affandar — COMMENTED** [review]
Left inline review comments for the MCP server PR.

File: packages/mcp-server/src/tools/agents.ts (hunk 1 above)
Comment (affandar) [reply packages/mcp-server/src/tools/agents.ts]: Once fixed, please do add a test for these as well.

File: packages/mcp-server/src/tools/models.ts (hunk 6 above)
Comment (affandar) [reply packages/mcp-server/src/tools/models.ts]: Add test for invalid commands sent as well so we could catch regressions in the future.

File: packages/mcp-server/src/tools/agents.ts (hunk 1 above)
Comment (ankitdas-volgapartners) [reply packages/mcp-server/src/tools/agents.ts]: Fixed in 97abd48. Dropped the `mgmt.sendCommand({ cmd: "spawn_agent" })` enqueue path entirely. `spawn_agent`, `message_agent`, and `cancel_agent` now use the real management APIs: `mgmt.createSessionForAgent` for spawn, cached-or-resumed `session.send()` (via `client.resumeSession(id)`) for message, and `mgmt.cancelSession` for cancel. No more silent `Unknown command` rejections in KV. Verified by LIVE smoke (`packages/mcp-server/test-mcp-b2-smoke.mjs`) hitting real Postgres + GitHub Copilot, 2/2 green back-to-back. Dedicated agent-tool tests still pending (tracked alongside the test-wiring item below).

File: packages/mcp-server/src/tools/models.ts (hunk 6 above)
Comment (ankitdas-volgapartners) [reply packages/mcp-server/src/tools/models.ts]: Fixed in 97abd48 via new `sendCommandAndWait` helper (`packages/mcp-server/src/util/command.ts`). Enqueues with a fresh cmdId, polls `mgmt.getCommandResponse` until response or timeout. `switch_model` and `send_command` now throw `CommandRejectedError` / `CommandTimeoutError` and surface them as MCP errors instead of returning `{ sent: true }`. Verified by unit test (`packages/mcp-server/test-sendCommandAndWait-unit.mjs`, 3/3 PASS) covering rejection, success, and timeout semantics with mocked mgmt. LIVE assertion deferred to unit level: `mgmt.sendCommand` requires orch `Started`, but `session.send()` returns before `Started`, so the cold-session race window is hard to hit deterministically from the smoke harness. The asymmetry between client (auto-boot) and mgmt (assumes live) APIs is partially deliberate per the comment block at `management-client.ts:1146-1182` (fails open on transient errors during boot), but the gap is real, happy to file an upstream issue tracking it separately.

Comment (ankitdas-volgapartners) [reply packages/mcp-server/src/tools/models.ts]: Done in 97abd48. `packages/mcp-server/test-sendCommandAndWait-unit.mjs` (3/3 PASS) explicitly covers the bogus-command rejection path: enqueue an unknown cmd, mgmt writes `error: "Unknown command: ..."` to KV, helper throws `CommandRejectedError`, caller surfaces it as MCP error. Also covers the success and timeout paths so all three branches are guarded against regressions. Wiring this into the package-level test script + CI is tracked with the broader test-wiring item.

**Comment by ankitdas-volgapartners** [comment]
Hi Affan,

Pushed 97abd48 covering cold-session send_message, bogus-command rejection, and the list_models field shape fix. Per-comment replies inline above.

Summary:
- **cold-session send_message**: routed through `client.resumeSession(id).send(msg)` instead of `mgmt.sendMessage(id, msg)`. Boots orch via `_startTurn` for cold sessions where the orch hasn't booted yet (previously threw "orchestration X is not started"). Same fix applied to `message_agent` for freshly-spawned children.
- **bogus-command rejection**: new `sendCommandAndWait` helper (`packages/mcp-server/src/util/command.ts`) enqueues with a fresh cmdId and polls `mgmt.getCommandResponse` until response or timeout. `switch_model` and `send_command` now surface `CommandRejectedError` / `CommandTimeoutError` as MCP errors instead of `{ sent: true }`.
- **list_models qualifiedName/modelName fix**: drops `(m: any).name` cast, emits real `qualifiedName` + `modelName` matching the SDK ModelDescriptor shape (prior cast silently produced `{ name: undefined, ... }`).
- **Agent-management tools**: same pattern, `spawn_agent` / `message_agent` / `cancel_agent` use real management APIs (`createSessionForAgent`, cached-or-resumed `session.send`, `mgmt.cancelSession`) instead of the unsupported `cmd: "spawn_agent"` enqueue.

Verified by:
- LIVE smoke (`packages/mcp-server/test-mcp-b2-smoke.mjs`) against real Postgres + GitHub Copilot, 2/2 green back-to-back. Covers cold-session send_message + list_models qualifiedName/modelName.
- Unit test (`packages/mcp-server/test-sendCommandAndWait-unit.mjs`) covering rejection / success / timeout semantics, 3/3 PASS.

One callout: bogus-command rejection sits at unit-test level rather than LIVE smoke. `mgmt.sendCommand` requires orch `Started`, `session.send()` returns before `Started`. The asymmetry between `client` (auto-boot) and `mgmt` (assumes live) APIs is partially deliberate per the comment block at `management-client.ts:1146-1182` (fails open on transient errors during boot), but the gap is real. Happy to file an upstream issue tracking it separately if useful.

Still pending on this PR: dumpSession projection, README npx form, dynamic SYSTEM_AGENTS, README counts, tests wiring + CI, hardening (files field, host flag, allowedHosts, constant-time auth, max-sessions, SIGTERM, dead auth.ts), N7 read_facts privilege doc/fix, lockfile rebase. Tackling next.

File: packages/mcp-server/src/resources/agents.ts (hunk 2 above)
Comment (ankitdas-volgapartners) [reply packages/mcp-server/src/resources/agents.ts]: Fixed in 97e9953. Hardcoded SYSTEM_AGENTS const dropped from `resources/agents.ts` and `resources/subscriptions.ts`. `ServerContext` now exposes `systemAgentIds: Set<string>`, populated once at startup via `mgmt.listSessions().filter(s => s.isSystem)` and refreshable mid-flight when subscription poller observes new ones. Resources iterate the dynamic set.

Side note: integration smoke logged 5 actual system agents (agent-tuner, facts-manager, pilotswarm, resourcemgr, sweeper) where the hardcoded list had 3, so the previous code was silently dropping resources for two real agents.

File: packages/mcp-server/src/resources/subscriptions.ts (hunk 3 above)
Comment (ankitdas-volgapartners) [reply packages/mcp-server/src/resources/subscriptions.ts]: Fixed in 97e9953. Hardcoded SYSTEM_AGENTS const dropped from `resources/agents.ts` and `resources/subscriptions.ts`. `ServerContext` now exposes `systemAgentIds: Set<string>`, populated once at startup via `mgmt.listSessions().filter(s => s.isSystem)` and refreshable mid-flight when subscription poller observes new ones. Resources iterate the dynamic set.

Side note: integration smoke logged 5 actual system agents (agent-tuner, facts-manager, pilotswarm, resourcemgr, sweeper) where the hardcoded list had 3, so the previous code was silently dropping resources for two real agents.

File: packages/mcp-server/README.md (hunk 4 above)
Comment (ankitdas-volgapartners) [reply packages/mcp-server/README.md]: Fixed in 438ef7a. README now uses `npx -y -p pilotswarm-mcp-server pilotswarm-mcp ...`, package name `pilotswarm-mcp-server` from package.json, bin name `pilotswarm-mcp`. Prior `npx pilotswarm-mcp ...` would fail because npm resolves the package name, not the bin.

File: packages/mcp-server/package.json (hunk 5 above)
Comment (ankitdas-volgapartners) [reply packages/mcp-server/package.json:23]: Fixed in ac061d1. Tests reorganized:

- `packages/mcp-server/test/unit/sendCommandAndWait.unit.mjs`, pure-mock, no DB/Copilot, runs in <2s
- `packages/mcp-server/test/integration/{smoke,tools,edge-cases,verify}.live.mjs`, LIVE suites gated on env

Package scripts: `npm test` (unit, CI-safe), `npm run test:integration` (smoke), `npm run test:integration:all` (full). Top-level aliases `test:mcp-server*` mirror these.

`scripts/run-tests.sh` now runs the unit suite as part of the standard pass (gated by `SKIP_MCP_SERVER_TESTS=1` if needed). CI workflow already invokes `scripts/run-tests.sh` after `npm run build --workspaces --if-present`, so the unit suite is picked up automatically. LIVE suite stays opt-in to avoid breaking CI when MCP-specific secrets aren't present.

File: packages/mcp-server/src/tools/agents.ts (hunk 1 above)
Comment (ankitdas-volgapartners) [reply packages/mcp-server/src/tools/agents.ts]: Partially addressed in 97abd48 + ac061d1. Agent-management tools (`spawn_agent`, `message_agent`, `cancel_agent`) now use real management APIs (`createSessionForAgent`, cached-or-resumed `session.send`, `mgmt.cancelSession`) and the test scaffolding under `packages/mcp-server/test/{unit,integration}/` is wired and ready. Dedicated unit + integration tests for the agent tools (asserting child session creation, message routing to child, cancel propagation) aren't in this PR yet, tracking separately as a follow-up.

**Comment by ankitdas-volgapartners** [comment]
Hi Affan,

Three follow-up commits since 97abd48:

- **`ac061d1`**, Tests reorganized into `packages/mcp-server/test/{unit,integration}/`. `npm test` runs the unit suite (CI-safe), `npm run test:integration` runs the LIVE smoke. `scripts/run-tests.sh` picks up the unit suite as part of the standard pass; existing CI workflow invokes it automatically. LIVE suite stays opt-in.

- **`97e9953`**, Hardcoded SYSTEM_AGENTS list dropped. `ServerContext.systemAgentIds` is populated dynamically at startup via `mgmt.listSessions().filter(s => s.isSystem)`. Integration smoke logged 5 actual system agents (agent-tuner, facts-manager, pilotswarm, resourcemgr, sweeper) where the hardcoded list had 3, so the previous code was silently dropping resources for two of them.

- **`438ef7a`**, README corrected: npx form now `npx -y -p pilotswarm-mcp-server pilotswarm-mcp`, tool count 15 to 20, resource count 5 to 15, discovery tools and per-system-agent resource URIs documented.

All commits verified by unit (3/3) + integration smoke (2/2) green.

Still pending: dumpSession projection (replace JSON.parse with getSessionEvents), read_facts caller-supplied session-id privilege doc/fix, hardening bundle (files field, host flag, allowedHosts, constant-time auth, max-sessions, SIGTERM, dead auth.ts), lockfile rebase, portal-API tracking issue. Tackling next.

**Comment by ankitdas-volgapartners** [comment]
Hi Affan,

Two more commits since the last batch:

- **`0173955`** — Fixed the `pilotswarm://sessions/{id}/messages` resource. It was using `JSON.parse(mgmt.dumpSession(id))` and `dumpSession()` returns Markdown, so the parse always threw and the catch silently returned `[]`. Switched to `mgmt.getSessionEvents(id, undefined, 100)`. Public URI / shape unchanged. Verified LIVE.

- **`fdd6bab`** — Hardening bundle: `files` field in package.json, `--host` flag (default 127.0.0.1), `--allowed-hosts` validation against the Host header, `crypto.timingSafeEqual` bearer compare, max-sessions cap (default 256, configurable via flag/env), SIGTERM + SIGINT graceful shutdown with 5s drain, and removal of the dead `src/auth.ts`. All manual checks pass (localhost-only bind, evil Host rejected, cap enforcement, clean SIGTERM exit). LIVE suite still green: unit 3/3, smoke 2/2.

Still pending: read_facts caller-supplied session-id privilege doc/fix, lockfile rebase against current main, portal API surface tracking issue. Tackling next.

File: packages/mcp-server/README.md (hunk 7 above)
Comment (ankitdas-volgapartners) [reply packages/mcp-server/README.md:23]: Tracker filed: https://github.com/affandar/PilotSwarm/issues/21. Covers the portal HTTP API surface design (mirroring `PilotSwarmManagementClient`), auth model for portal-facing clients, a portal-mode transport for the MCP server (so it can be configured with portal URL + auth instead of `DATABASE_URL`), versioning and deprecation policy, TUI migration path, and CI contract tests.

Also tightened the trust posture in the meantime (commit `0bccb97`): `read_facts` and other caller-supplied-`session_id` tools now have an explicit TRUST NOTE in their tool description, and the README has a new `## Security model` section calling out the caller-trusted semantics, plus what the server does enforce (Host allowlist, `crypto.timingSafeEqual` bearer compare, max-sessions cap, CORS). Stricter per-client session ownership belongs at the portal/transport layer once that surface lands.

File: package-lock.json (hunk 8 above)
Comment (ankitdas-volgapartners) [reply package-lock.json:3654]: Resolved earlier in merge commit `d7ff0ad`. Verified current state: branch is 11 ahead, 0 behind `upstream/main`, merge-base equals `upstream/main` HEAD (`1398a8d`), no new lockfile conflict since the merge.

**Comment by ankitdas-volgapartners** [comment]
Hi Affan,

Two more commits closing out this batch:

- **`0173955`** — Resource fix on `pilotswarm://sessions/{id}/messages`. Was using `JSON.parse(mgmt.dumpSession(id))` and `dumpSession()` returns Markdown, so parse always threw and the catch silently returned `[]`. Switched to `mgmt.getSessionEvents(id, undefined, 100)`. Public URI / shape unchanged. Verified LIVE.

- **`fdd6bab`** — Hardening bundle: `files` field in package.json, `--host` flag (default 127.0.0.1), `--allowed-hosts` validation against the Host header, `crypto.timingSafeEqual` bearer compare, max-sessions cap (default 256, configurable), SIGTERM + SIGINT graceful shutdown with 5s drain, removal of dead `src/auth.ts`. Manual checks all pass.

- **`0bccb97`** — `read_facts` privilege boundary documented: code-level TRUST NOTE in `tools/facts.ts`, new `## Security model` section in README. Tracker issue for portal API surface filed at https://github.com/affandar/PilotSwarm/issues/21.

Full commit list since branch open: 97abd48 (cold-session send_message + bogus-command rejection + list_models field shape), ac061d1 (test reorganization + scripts/CI wiring), 97e9953 (dynamic SYSTEM_AGENTS via management API), 438ef7a (README: npx form, tool/resource counts, discovery tools), 0173955 (sessions/{id}/messages resource fix), fdd6bab (hardening bundle), 0bccb97 (read_facts privilege docs).

Branch state: 11 ahead, 0 behind upstream/main, lockfile current.

**Comment by ankitdas-volgapartners** [comment]
Hi Affan,

Ran a fresh cold review and full functional QA pass before pinging you. Found and fixed several real defects. 5 new commits since the last batch:

**`17f845a`** , `fix(mcp-server): validate session_id and gate session-scoped tools on existence`. The session-scoped tools (`send_message`, `send_and_wait`, `send_answer`, `abort_session`, `rename_session`, `get_session_detail`, `delete_session`, `get_session_events`) were returning `{sent: true}` / etc. for fake or malformed `session_id` values (`"not-a-uuid"`, all-zero UUIDs). `client.resumeSession()` then enqueued worker entries that created phantom orchestration rows in `duroxide.instances`. Schema is now `z.string().uuid()` and every handler runs an `mgmt.getSession()` pre-flight check returning a typed not-found error. Folded into the same commit: `create_session(agent: ..., system_message: ...)` was silently dropping `systemMessage` because `client.createSessionForAgent` doesn't accept it; routes through `client.createSession` directly now. The `pilotswarm://sessions/{id}` resource returned the literal string `"null"` for missing ids; now returns a structured error body. `title` bounded `min(1).max(512)` on create and rename.

**`5e6667c`** , `fix(mcp-server): pilotswarm://models resource emits qualified_name + model_name`. Same `(m: any).name` cast bug 97abd48 fixed in `list_models`, missed in the resource path. SDK ModelDescriptor has no `name` field; resource was emitting `name: undefined` which JSON.stringify strips entirely. Drops the casts, emits real `{qualified_name, model_name, provider, description, cost}`.

**`f605b06`** , `fix(mcp-server): declare MCP capabilities and register subscribe handlers with Zod schemas`. Two bugs made resource subscriptions and `prompts/list` unreachable from clients. `new McpServer(...)` was constructed without a capabilities argument, so default-config (no `--plugin`) advertised neither `prompts` nor `resources.subscribe: true`. And `resources/subscriptions.ts` was calling `setRequestHandler({method: "resources/subscribe"}, ...)` with a plain object where the SDK requires a Zod schema; the throw was wrapped in a silent try/catch so the 190-line poller never installed. Both fixed; verified live, `initialize` now advertises `{prompts, resources: {subscribe: true, listChanged: true}, tools}`, `resources/subscribe` returns `{}`, `prompts/list` returns `{prompts: []}`.

**`da2bbdb`** , `chore(mcp-server): HTTP idle-session sweeper, atomic max-sessions cap, log-level wiring, README CLI docs`. HTTP one-shot POSTs leaked their max-sessions slot indefinitely because `transport.onclose` doesn't fire for short-lived clients. With `--max-sessions 4`, four curl one-shots permanently locked the cap. Added an idle-session sweeper (default 5 min, configurable). The cap check itself had a TOCTOU race that a concurrent burst could exploit; replaced with an atomic counter incremented before transport construction. `--log-level` was parsed and never read; wired to a tier-based log helper. README CLI Options table refreshed with `--host`, `--allowed-hosts`, `--max-sessions`, `--session-idle-timeout-ms`, plus the `PILOTSWARM_MCP_*` env aliases.

**`79cc908`** , `fix(mcp-server): spawn_agent links child to parent and forwards system_message`. `spawn_agent` was passing `parentSessionId` and `systemMessage` to `client.createSessionForAgent`, but the SDK signature accepts neither. Both fields were silently dropped by TypeScript structural widening, so spawned children had `parentSessionId = null` in CMS, became peer top-level sessions, and never appeared in `list_agents`. Routed through `client.createSession({agentId, boundAgentName, promptLayering: {kind: "app-agent"}, parentSessionId, ...})`. Setting `parentSessionId` triggers `isSubAgent === true` which bypasses the allowlist check, so no context-side widening of `allowedAgentNames` was needed. Verified live: DB row shows the parent linkage, `list_agents({parent_session_id})` returns the child. One known cosmetic gap documented inline: `client.createSession` does not write `agentId` to the CMS catalog row, so spawned children show `agent_id: null` in `list_agents` rows even though the orchestration receives `agentId` correctly. Fix requires an SDK change.

Verification across the round: unit 3/3, integration smoke 2/2, tools live 17/25 PASS (up from 16; `delete_session` now passes once an orchestration row exists), edge-cases unchanged. Manual stdio + HTTP exercises confirm the protocol-level fixes and the slot-leak / cap / spawn_agent / models-resource fixes end-to-end.

Still pending: tightening existing test assertions. Several PASS gates accept too-broad outputs (some return `[]` and pass, one accepts both clean-error and unexpected-success), and two tests pass wrong arg names. Coming as a follow-up commit so the suite can catch regressions of the bugs in this round.

**Comment by ankitdas-volgapartners** [comment]
Appended 5 commits on top of `79cc908`:

1. `a448784 fix(mcp-server): delete_session falls back to client.deleteSession when orchestration NotFound` — clears stale rows whose orchestration was never started.
2. `14bb92e feat(mcp-server): create_session refuses + auto-purges row when no worker claims orchestration` — guard polls `mgmt.getSessionStatus` for up to `PILOTSWARM_MCP_WORKER_CLAIM_TIMEOUT_MS` (default 10s); on timeout returns `{error:"no_worker_claimed", created:false, purged_session_id, message}` and rolls back the CMS row.
3. `d2c0457 refactor(mcp-server): drop spawn_agent / message_agent / cancel_agent from external surface` — addresses the review: external MCP mutates only top-level sessions; sub-agent lifecycle belongs to the parent's in-loop reasoning via `spawn_agent` / orchestration commands.
4. `4a3506d feat(mcp-server): add list_registered_agents / get_agent_tree / get_session_tree_stats inspection tools` — fills the read-side gap left by the removals. `list_registered_agents` reads the catalog; `get_agent_tree` does in-memory BFS over `parent_session_id` with cycle defense and configurable depth; `get_session_tree_stats` snake-cases `mgmt.getSessionTreeStats`.
5. `d4e37e5 docs(mcp-server): document external MCP sub-agent boundary + refresh tool table` — README "External MCP boundary" section + tool-table refresh.

**Net surface change:** −3 mutating tools, +3 read-only tools.

**No SDK changes.** Build clean (`npm run build --workspace=pilotswarm-mcp-server`).
